From 7344f396d1465c22dd656e4eb1c9f3117079d0c7 Mon Sep 17 00:00:00 2001
From: daikai <daikai@uniontech.com>
Date: Wed, 5 Aug 2020 09:40:52 +0800
Subject: [PATCH] fix(concheck):concheck takes too long time

It takes too long to check connectivity when there are multiple network cards.Some devices are not connected, but still waiting for http results.
---
 src/devices/nm-device.c | 1 +
 src/nm-connectivity.c   | 4 +++-
 src/nm-connectivity.h   | 1 +
 3 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/devices/nm-device.c b/src/devices/nm-device.c
index 89e1cc517..37938967e 100644
--- a/src/devices/nm-device.c
+++ b/src/devices/nm-device.c
@@ -2975,6 +2975,7 @@ concheck_start (NMDevice *self,
 	       is_periodic ? ", periodic-check" : "");
 
 	handle->c_handle = nm_connectivity_check_start (concheck_get_mgr (self),
+	                                                nm_device_get_state (self),
 	                                                nm_device_get_ip_iface (self),
 	                                                concheck_cb,
 	                                                handle);
diff --git a/src/nm-connectivity.c b/src/nm-connectivity.c
index 8a6e955ae..0b0239044 100644
--- a/src/nm-connectivity.c
+++ b/src/nm-connectivity.c
@@ -558,6 +558,7 @@ _idle_cb (gpointer user_data)
 
 NMConnectivityCheckHandle *
 nm_connectivity_check_start (NMConnectivity *self,
+                             NMDeviceState nmstate,
                              const char *iface,
                              NMConnectivityCheckCallback callback,
                              gpointer user_data)
@@ -578,11 +579,12 @@ nm_connectivity_check_start (NMConnectivity *self,
 	cb_data->user_data = user_data;
 	cb_data->completed_state = NM_CONNECTIVITY_UNKNOWN;
 
+	_LOG2D ("nm_connectivity_check_start( '%s',%d)", iface,nmstate);
 	if (iface)
 		cb_data->ifspec = g_strdup_printf ("if!%s", iface);
 
 #if WITH_CONCHECK
-	if (iface) {
+	if (iface  && nmstate == NM_DEVICE_STATE_ACTIVATED) {
 		CURL *ehandle;
 
 		if (   priv->enabled
diff --git a/src/nm-connectivity.h b/src/nm-connectivity.h
index 99333cede..68018b74f 100644
--- a/src/nm-connectivity.h
+++ b/src/nm-connectivity.h
@@ -73,6 +73,7 @@ typedef void (*NMConnectivityCheckCallback) (NMConnectivity *self,
                                              gpointer user_data);
 
 NMConnectivityCheckHandle *nm_connectivity_check_start (NMConnectivity *self,
+                                                        NMDeviceState nmstate,
                                                         const char *iface,
                                                         NMConnectivityCheckCallback callback,
                                                         gpointer user_data);
-- 
2.20.1

